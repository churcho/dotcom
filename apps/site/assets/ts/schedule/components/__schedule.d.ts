// import {  } from "../../__v3api";

export interface SchedulePageData {
  pdfs: SchedulePDF[];
  teasers: string | null;
  hours: string;
}

export interface SchedulePDF {
  title: string;
  url: string;
}
