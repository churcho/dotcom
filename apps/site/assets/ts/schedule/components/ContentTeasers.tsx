import React, { ReactElement } from "react";

interface Props {
  teasers: string | null;
}

const ContentTeasers = ({ teasers }: Props): ReactElement<HTMLElement> | null =>
  teasers ? <div dangerouslySetInnerHTML={{ __html: teasers }} /> : null;

export default ContentTeasers;
