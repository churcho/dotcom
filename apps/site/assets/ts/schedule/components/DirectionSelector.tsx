import React, { ReactElement } from "react";
import SelectContainer from "./schedule-finder/SelectContainer";
import { RoutePattern, DirectionId, Route } from "../../__v3api";

interface Props {
  directionId: DirectionId;
  route: Route;
  routePatterns: RoutePattern[];
}

const DirectionSelector = ({
  directionId,
  route,
  routePatterns
}: Props): ReactElement<HTMLElement> => {
  const directionName: string = route.direction_names[directionId];
  const directionDestination: string =
    route.direction_destinations[directionId];
  return (
    <div>
      <div className="u-small-caps" style={{ fontWeight: "bold" }}>
        {directionName}
      </div>
      <h3 style={{ marginTop: 0 }}>{directionDestination}</h3>
      {routePatterns.length > 1 && (
        <SelectContainer id="direction-selector" error={false}>
          <select>
            {routePatterns.map(pattern => (
              <option value={pattern.id}>
                {pattern.name} {pattern.typicality}{" "}
                {pattern.time_desc && `(${pattern.time_desc})`}
              </option>
            ))}
          </select>
        </SelectContainer>
      )}
    </div>
  );
};

export default DirectionSelector;
