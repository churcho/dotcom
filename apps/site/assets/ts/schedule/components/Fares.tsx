import React, { ReactElement } from "react";
import ExpandableBlock from "../../components/ExpandableBlock";
import { Fare } from "./__schedule";

const fareItem = (fare: Fare): ReactElement<HTMLElement> => (
  <p key={fare.title} className="m-schedule-page__fare">
    <span>{fare.title}</span>
    <span>{fare.price}</span>
  </p>
);

interface Props {
  fares: Fare[];
  fareLink: string;
}

const Fares = ({ fares, fareLink }: Props): ReactElement<HTMLElement> | null =>
  fares.length > 0 ? (
    <ExpandableBlock
      header={{ text: "Fares", iconSvgText: null }}
      initiallyExpanded={false}
      id="fares"
    >
      <>
        {fares.map(f => fareItem(f))}
        <p className="m-schedule-page__link">
          <a className="c-call-to-action" href={fareLink}>
            More about fares
          </a>
        </p>
      </>
    </ExpandableBlock>
  ) : null;

export default Fares;
