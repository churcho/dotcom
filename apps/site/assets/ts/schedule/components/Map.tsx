import React, { ReactElement } from "react";
import { LatLng, latLng, LatLngBounds, latLngBounds } from "leaflet";
import Map from "../../leaflet/components/Map";
import {
  MapData,
  MapMarker as Marker
} from "../../leaflet/components/__mapdata";

interface Props {
  data: MapData;
}

export const getBounds = (markers: Marker[]): LatLngBounds => {
  const points: LatLng[] = markers.map(m => latLng(m.latitude, m.longitude));
  return latLngBounds(points);
};

export const iconOpts = (
  icon: string | null
): {
  iconSize?: [number, number];
  iconAnchor?: [number, number];
} => {
  switch (icon) {
    case null:
      return {};

    case "vehicle-bordered-expanded":
      return {
        iconSize: [18, 18],
        iconAnchor: [6, 6]
      };

    case "stop-circle-bordered-expanded":
      return {
        iconSize: [12, 12],
        iconAnchor: [6, 6]
      };

    default:
      throw new Error(`unexpected icon type: ${icon}`);
  }
};

const zIndex = (icon: string | null): number | undefined =>
  icon === "vehicle-bordered-expanded" ? 1000 : undefined;

const updateMarker = (marker: Marker): Marker => ({
  ...marker,
  tooltip: <div>{marker.tooltip_text}</div>,
  iconOpts: iconOpts(marker.icon),
  zIndex: zIndex(marker.icon)
});

export default ({ data }: Props): ReactElement<HTMLElement> => (
  <div className="m-schedule__map">
    <Map
      bounds={getBounds(data.markers)}
      mapData={{ ...data, markers: data.markers.map(updateMarker) }}
    />
  </div>
);
