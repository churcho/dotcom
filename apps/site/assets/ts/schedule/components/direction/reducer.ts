import { DirectionId, Shape } from "../../../__v3api";
import {
  ShapesById,
  RoutePatternsByDirection,
  EnhancedRoutePattern
} from "../__schedule";

export interface State {
  routePattern: EnhancedRoutePattern;
  shape: Shape;
  directionId: DirectionId;
  shapesById: ShapesById;
  routeId: string;
  routePatternsByDirection: RoutePatternsByDirection;
  routePatternMenuOpen: boolean;
  routePatternMenuAll: boolean;
  itemFocus: string | null;
  stopListHtml: string;
  firstStopListRender: boolean;
}

export interface Payload {
  routePattern?: EnhancedRoutePattern;
  newStopListHtml?: string;
}

export interface Action {
  event:
    | "toggleDirection"
    | "setRoutePattern"
    | "toggleRoutePatternMenu"
    | "closeRoutePatternMenu"
    | "showAllRoutePatterns"
    | "firstStopListRender"
    | "newStopListHtml";
  payload?: Payload;
}

const toggleDirection = (state: State): State => {
  const nextDirection = state.directionId === 0 ? 1 : 0;
  const [defaultRoutePatternForDirection] = state.routePatternsByDirection[
    nextDirection
  ];

  return {
    ...state,
    directionId: nextDirection,
    routePattern: defaultRoutePatternForDirection,
    shape: state.shapesById[defaultRoutePatternForDirection.shape_id],
    itemFocus: "first"
  };
};

const toggleRoutePatternMenu = (state: State): State => ({
  ...state,
  routePatternMenuOpen: !state.routePatternMenuOpen,
  itemFocus: "first"
});

const showAllRoutePatterns = (state: State): State => ({
  ...state,
  routePatternMenuAll: true,
  itemFocus: "first-uncommon"
});

export const reducer = (state: State, action: Action): State => {
  switch (action.event) {
    case "toggleDirection":
      return toggleDirection(state);

    case "toggleRoutePatternMenu":
      return toggleRoutePatternMenu(state);

    case "showAllRoutePatterns":
      return showAllRoutePatterns(state);

    case "closeRoutePatternMenu":
      return {
        ...state,
        routePatternMenuOpen: false,
        itemFocus: null
      };

    case "setRoutePattern":
      return {
        ...state,
        routePattern: action.payload!.routePattern!,
        shape: state.shapesById[action.payload!.routePattern!.shape_id],
        routePatternMenuOpen: false,
        itemFocus: null
      };

    case "firstStopListRender":
      return { ...state, firstStopListRender: false };

    case "newStopListHtml":
      return { ...state, stopListHtml: action.payload!.newStopListHtml! };

    /* istanbul ignore next */
    default:
      return state;
  }
};
