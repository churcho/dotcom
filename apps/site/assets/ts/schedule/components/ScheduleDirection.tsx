import React, { Dispatch, ReactElement } from "react";
import { DirectionId, EnhancedRoute } from "../../__v3api";
import { ShapesById, RoutePatternsByDirection } from "./__schedule";
import ScheduleDirectionMenu from "./direction/ScheduleDirectionMenu";
import ScheduleDirectionButton from "./direction/ScheduleDirectionButton";
import { Action, State } from "./direction/reducer";

export interface Props {
  route: EnhancedRoute;
  directionId: DirectionId;
  shapesById: ShapesById;
  routePatternsByDirection: RoutePatternsByDirection;
  state: State;
  dispatch: Dispatch<Action>;
}

const ScheduleDirection = ({
  route,
  directionId,
  shapesById,
  routePatternsByDirection,
  state,
  dispatch
}: Props): ReactElement<HTMLElement> => {
  return (
    <div className="m-schedule-direction">
      <div id="direction-name" className="m-schedule-direction__direction">
        {route.direction_names[state.directionId]}
      </div>
      <ScheduleDirectionMenu
        route={route}
        directionId={state.directionId}
        routePatternsByDirection={routePatternsByDirection}
        selectedRoutePatternId={state.routePattern.id}
        menuOpen={state.routePatternMenuOpen}
        showAllRoutePatterns={state.routePatternMenuAll}
        itemFocus={state.itemFocus}
        dispatch={dispatch}
      />
      <ScheduleDirectionButton dispatch={dispatch} />
    </div>
  );
};

export default ScheduleDirection;
