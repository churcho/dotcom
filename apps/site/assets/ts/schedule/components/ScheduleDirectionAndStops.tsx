import React, { ReactElement, useReducer, useState } from "react";
import ScheduleDirection from "./ScheduleDirection";
import { SchedulePageData } from "./__schedule";
import { DirectionId } from "../../__v3api";
import { reducer } from "./direction/reducer";

export interface SchedulePageDataWithStopListContent extends SchedulePageData {
  stop_list_html: string;
}

const ScheduleDirectionAndStops = (schedulePageData: SchedulePageDataWithStopListContent): ReactElement<HTMLElement> => {
  const {
    direction_id: directionId,
    route,
    shape_map: shapesById,
    route_patterns: routePatternsByDirection,
  } = schedulePageData;

  const defaultRoutePattern = routePatternsByDirection[directionId].slice(
    0,
    1
  )[0];

  const [stopListHtml, updateStopListHtml] = useState(schedulePageData.stop_list_html);
  const [state, dispatch] = useReducer(reducer, {
    routePattern: defaultRoutePattern,
    shape: shapesById[defaultRoutePattern.shape_id],
    directionId,
    shapesById,
    routePatternsByDirection,
    routePatternMenuOpen: false,
    routePatternMenuAll: false,
    itemFocus: null
  });


  return (
    <>
      <ScheduleDirection
        directionId={schedulePageData.direction_id}
        route={schedulePageData.route}
        shapesById={schedulePageData.shape_map}
        routePatternsByDirection={schedulePageData.route_patterns}
        state={state}
        dispatch={dispatch}
      />
      <div dangerouslySetInnerHTML={{__html: stopListHtml}} />
    </>
  );
}

export default ScheduleDirectionAndStops;
