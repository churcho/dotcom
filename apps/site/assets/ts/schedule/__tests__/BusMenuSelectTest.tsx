import React from "react";
import renderer, { act } from "react-test-renderer";
import { BusMenuSelect } from "../components/direction/BusMenu";
import { EnhancedRoutePattern } from "../components/__schedule";

describe("BusMenuSelect", () => {
  it("renders properly for the Silver Line Waterfront", () => {
    const routePatterns: EnhancedRoutePattern[] = [
      {
        typicality: 1,
        time_desc: null,
        shape_id: "7460006",
        route_id: "746",
        representative_trip_id: "43160892",
        name: "Silver Line Way - South Station",
        id: "746-_-1",
        headsign: "South Station",
        direction_id: 1
      }
    ];
    let tree;
    act(() => {
      tree = renderer.create(
        <BusMenuSelect
          clickableMenu={false}
          routePatterns={routePatterns}
          selectedRoutePatternId={"746-_-1"}
          dispatch={() => {}}
        />
      );
    });
    expect(tree).toMatchSnapshot();
  });
});
