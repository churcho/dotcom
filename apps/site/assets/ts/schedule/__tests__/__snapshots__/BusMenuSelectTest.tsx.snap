// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusMenuSelect renders properly for the Silver Line Waterfront 1`] = `
<div
  className="m-schedule-direction__route-pattern"
  onClick={[Function]}
  onKeyUp={[Function]}
>
  Silver Line Waterfront
   
</div>
`;
